crates/store/re_video/src/decode/ffmpeg.rs

```diff
@@ -0,0 +1,730 @@
+                    format,
+                    pix_fmt,
+                    width,
+                    height,
+                    fps,
+                    ..
+                } = stream;
+                re_log::trace!(
+                    "{debug_name} decoder output: {stream_type} {format} {pix_fmt} {width}x{height} @ {fps} FPS"
+                );
+
+                debug_assert_eq!(stream_type.to_ascii_lowercase(), "video");
+            }
+
+            FfmpegEvent::Progress(_) => {
+                // We can get out frame number etc here to know how far behind we are.
+                // By default this triggers every 5s.
+            }
+
+            FfmpegEvent::OutputFrame(frame) => {
+                // Frames come in PTS order, but "frame info" comes in DTS order!
+                //
+                // Whenever the highest known DTS is behind the PTS, we need to wait until the DTS catches up.
+                // Otherwise, we'd assign the wrong PTS to the frame that just came in.
+                //
+                // Example how presentation timestamps and decode timestamps
+                // can play out in the presence of B-frames to illustrate this:
+                //    PTS: 1 4 2 3
+                //    DTS: 1 2 3 4
+                // Stream: I P B B
+                let frame_info = loop {
+                    if pending_frame_infos
+                        .first_key_value()
+                        .map_or(true, |(pts, _)| *pts > highest_dts)
+                    {
+                        let Ok(frame_info) = frame_info_rx.try_recv() else {
+                            re_log::debug!(
+                                "{debug_name} ffmpeg decoder frame info channel disconnected"
+                            );
+                            return;
+                        };
+
+                        // If the decodetimestamp did not increase, we're probably seeking backwards!
+                        // We'd expect the video player to do a reset prior to that and close the channel as part of that, but we may not have noticed that in here yet!
+                        // In any case, we'll have to just run with this as the new highest timestamp, not much else we can do.
+                        if highest_dts < frame_info.decode_timestamp {
+                            re_log::warn_once!("Video decode timestamps are expected to monotonically increase unless there was a decoder reset. This is probably a bug in Rerun.");
+                        }
+                        highest_dts = frame_info.decode_timestamp;
+
+                        pending_frame_infos.insert(frame_info.presentation_timestamp, frame_info);
+                    } else {
+                        // There must be an element here, otherwise we wouldn't be in this branch.
+                        #[allow(clippy::unwrap_used)]
+                        break pending_frame_infos.pop_first().unwrap().1;
+                    }
```

Comment:
I have a hard time following this logic. Does this work?

```suggestion
                    let oldest_pts_in_buffer = pending_frame_infos.first_key_value().map(|(pts, _)| *pts);
                    let is_caught_up = oldest_pts_in_buffer.is_some_and(|pts| pts <= highest_dts);
                    if is_caught_up {
                          // There must be an element here, otherwise we wouldn't be here.
                          #[allow(clippy::unwrap_used)] 
                          break pending_frame_infos.pop_first().unwrap().1;
                    } else {
                        // We're behind:
	                    let Ok(frame_info) = frame_info_rx.try_recv() else {
	                        re_log::debug!(
	                            "{debug_name} ffmpeg decoder frame info channel disconnected"
	                        );
	                        return;
	                    };
	
	                    // If the decodetimestamp did not increase, we're probably seeking backwards!
	                    // We'd expect the video player to do a reset prior to that and close the channel as part of that, but we may not have noticed that in here yet!
	                    // In any case, we'll have to just run with this as the new highest timestamp, not much else we can do.
	                    if highest_dts < frame_info.decode_timestamp {
	                        re_log::warn_once!("Video decode timestamps are expected to monotonically increase unless there was a decoder reset. This is probably a bug in Rerun.");
	                    }
	                    highest_dts = frame_info.decode_timestamp;
	
	                    pending_frame_infos.insert(frame_info.presentation_timestamp, frame_info);
	                }
```